<?xml version="1.0" ?>
<robot name="rover_18">
  <gazebo>
    <self_collide>true</self_collide>
  </gazebo>

  <!--Links defined below-->  
  <link name="base_link">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="2.7918" />
      <inertia ixx="0.095684"
               ixy="0.001902"
               ixz="-0.00027184"
               iyy="0.058185"
               iyz="0.00031126"
               izz="0.14455" />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/base_link.stl" />
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/base_link.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="base_link">
    <material>Gazebo/White</material>
  </gazebo>

  <link name="kinect">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="2.7918" />
      <inertia ixx="0.095684"
               ixy="0.001902"
               ixz="-0.00027184"
               iyy="0.058185"
               iyz="0.00031126"
               izz="0.14455" />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.1 0.6 0.1"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.2 0.05 0.05"/>
      </geometry>
    </visual>
  </link>
  <gazebo reference="kinect">
    <material>Gazebo/Black</material>
  </gazebo>

  <link name="wheel1">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="2.7918" />
      <inertia ixx="0.095684"
               ixy="0.001902"
               ixz="-0.00027184"
               iyy="0.058185"
               iyz="0.00031126"
               izz="0.14455" />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel1.stl" />
      </geometry>
      <surface>
        <bounce>
          <restitution_coefficient>0.1</restitution_coefficient>
          <threshold>0</threshold>
        </bounce>
        <contact>
          <ode>
            <max_vel>0.2</max_vel>
          </ode>
        </contact>
      </surface>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel1.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="wheel1">
    <material>Gazebo/DarkGrey</material>
  </gazebo>

  <link name="wheel2">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 3.14" />
      <mass value="2.7918" />
      <inertia ixx="0.095684"
               ixy="0.001902"
               ixz="-0.00027184"
               iyy="0.058185"
               iyz="0.00031126"
               izz="0.14455" />    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 3.14" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel2.stl" />
      </geometry>
      <surface>
        <bounce>
          <restitution_coefficient>0.1</restitution_coefficient>
          <threshold>0</threshold>
        </bounce>
        <contact>
          <ode>
            <max_vel>0.2</max_vel>
          </ode>
        </contact>
      </surface>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 3.14" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel2.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="wheel2">
    <material>Gazebo/DarkGrey</material>
  </gazebo>

  <link name="wheel3">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="2.7918" />
      <inertia ixx="0.095684"
               ixy="0.001902"
               ixz="-0.00027184"
               iyy="0.058185"
               iyz="0.00031126"
               izz="0.14455" />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel3.stl" />
      </geometry>
      <surface>
        <bounce>
          <restitution_coefficient>0.1</restitution_coefficient>
          <threshold>0</threshold>
        </bounce>
        <contact>
          <ode>
            <max_vel>0.2</max_vel>
          </ode>
        </contact>
      </surface>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel3.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="wheel3">
    <material>Gazebo/DarkGrey</material>
  </gazebo>

    <link name="wheel4">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 3.14" />
      <mass value="2.7918" />
      <inertia ixx="0.095684"
               ixy="0.001902"
               ixz="-0.00027184"
               iyy="0.058185"
               iyz="0.00031126"
               izz="0.14455" />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 3.14" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel4.stl" />
      </geometry>
      <surface>
        <bounce>
          <restitution_coefficient>0.1</restitution_coefficient>
          <threshold>0</threshold>
        </bounce>
        <contact>
          <ode>
            <max_vel>0.2</max_vel>
          </ode>
        </contact>
      </surface>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 3.14" />
      <geometry>
        <mesh filename="package://rover_18/meshes/wheel4.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="wheel4">
    <material>Gazebo/DarkGrey</material>
  </gazebo>

  <link name="base_rotation">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="0.773197125339031" />
      <inertia ixx="0.00103764705393415"
               ixy="6.58182970956898E-06"
               ixz="5.54891502901135E-05"
               iyy="0.00130367193834491"
               iyz="-0.000179327606683225"
               izz="0.000348493395648461" />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/base_rotation.stl" />
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/base_rotation.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="base_rotation">
    <material>Gazebo/Grey</material>
  </gazebo>

  <link name="shoulder_links">
    <inertial>
      <origin xyz="-0.005 0.095 -0.12" rpy="0 0 0" />
      <mass value="0.104640684850347" />
      <inertia ixx="7.98235631104724E-05"
               ixy="-4.92478395888344E-19"
               ixz="4.07842412768808E-05"
               iyy="0.000925048572837215"
               iyz="-3.03056955059653E-18"
               izz="0.000973974480713755" />
    </inertial>
    <collision>
      <origin xyz="-0.005 0.095 -0.12" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/shoulder.stl" />
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.005 0.095 -0.12" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/shoulder.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="shoulder_links">
    <material>Gazebo/Grey</material>
  </gazebo>

  <link name="elbow_link">
    <inertial>
      <origin xyz="0 0.18 -0.02" rpy="0 0 0" />
      <mass value="0.0411062379032747" />
      <inertia ixx="1.21837641589144E-05"
               ixy="-1.69964129768756E-18"
               ixz="1.52307939635103E-06"
               iyy="0.000132211701064061"
               iyz="5.81115462912697E-19"
               izz="0.00013982909112128" />
    </inertial>
    <collision>
      <origin xyz="0 0.18 -0.02" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/elbow.stl" />
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0.18 -0.02" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/elbow.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="elbow_link">
    <material>Gazebo/Grey</material>
  </gazebo>

  <link name="gripper">
    <inertial>
      <origin xyz="0.005 0.02 -0.18" rpy="0 0 0" />
      <mass value="0.0411062379032747" />
      <inertia ixx="1.21837641589144E-05"
               ixy="-1.69964129768756E-18"
               ixz="1.52307939635103E-06"
               iyy="0.000132211701064061"
               iyz="5.81115462912697E-19"
               izz="0.00013982909112128" />
    </inertial>
    <collision>
      <origin xyz="0.005 0.02 -0.18" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/gripper.stl" />
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.005 0.02 -0.18" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://rover_18/meshes/gripper.stl" />
      </geometry>
    </visual>
  </link>
  <gazebo reference="gripper">
    <material>Gazebo/Grey</material>
  </gazebo>


  <!--Joints defined below  **LIMIT REVIEW TO BE DONE** -->  
  <joint name="kinect_joint" type="fixed">
    <origin xyz="0 0.35 0.12" rpy="-1.5 0 0" />
    <parent link="base_link" />
    <child link="kinect" />
  </joint>

  <joint name="wheel1_joint" type="continuous">
    <origin xyz="0.46 0.43 -0.13" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="wheel1" />
    <axis xyz="1 0 0" />
<!--     <limit effort="200" velocity="1" /> -->
  </joint>
  <gazebo reference="wheel1_joint">
    <sensor name="wheel1_ft_sensor" type="force_torque">
      <always_on>true</always_on>
      <update_rate>1.0</update_rate>
      <force_torque>
        <frame>parent</frame>
        <measure_direction>child_to_parent</measure_direction>
      </force_torque>
    </sensor>
  </gazebo>

  <transmission name="$wheel1_trans">
    <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$wheel1_motor">
        <hardwareInterface>EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="wheel1_joint">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <joint name="wheel2_joint" type="continuous">
    <origin xyz="-0.46 0.44 -0.13" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="wheel2" />
    <axis xyz="1 0 0" />
<!--     <limit effort="200" velocity="1" /> -->
  </joint>
  <gazebo reference="wheel2_joint">
    <sensor name="wheel2_ft_sensor" type="force_torque">
      <always_on>true</always_on>
      <update_rate>1.0</update_rate>
      <force_torque>
        <frame>parent</frame>
        <measure_direction>child_to_parent</measure_direction>
      </force_torque>
    </sensor>
  </gazebo>

  <transmission name="$wheel2_trans">
    <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$wheel2_motor">
        <hardwareInterface>EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="wheel2_joint">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <joint name="wheel3_joint" type="continuous">
    <origin xyz="0.46 -0.45 -0.13" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="wheel3" />
    <axis xyz="1 0 0" />
<!--     <limit effort="200" velocity="1" /> -->
  </joint>
  <gazebo reference="wheel3_joint">
    <sensor name="wheel3_ft_sensor" type="force_torque">
      <always_on>true</always_on>
      <update_rate>1.0</update_rate>
      <force_torque>
        <frame>parent</frame>
        <measure_direction>child_to_parent</measure_direction>
      </force_torque>
    </sensor>
  </gazebo> 

  <transmission name="$wheel3_trans">
    <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$wheel3_motor">
        <hardwareInterface>EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="wheel3_joint">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
  </transmission> 

  <joint name="wheel4_joint" type="continuous">
    <origin xyz="-0.46 -0.44 -0.13" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="wheel4" />
    <axis xyz="1 0 0" />
<!--     <limit effort="200" velocity="1" /> -->
  </joint>
    <gazebo reference="wheel4_joint">
    <sensor name="wheel4_ft_sensor" type="force_torque">
      <always_on>true</always_on>
      <update_rate>1.0</update_rate>
      <force_torque>
        <frame>parent</frame>
        <measure_direction>child_to_parent</measure_direction>
      </force_torque>
    </sensor>
  </gazebo>

  <transmission name="$wheel4_trans">
    <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$wheel4_motor">
        <hardwareInterface>EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="wheel4_joint">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
  </transmission>  

  <joint name="base_rotation_joint" type="fixed">
    <origin xyz="0 0.19 0.14" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="base_rotation" />
  </joint>

  <joint name="shoulder_joint" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <parent link="base_rotation" />
    <child link="shoulder_links" />
  </joint>

  <joint name="elbow_joint" type="fixed">
    <origin xyz="-0.01 0.225 0.515" rpy="0 0 0" />
    <parent link="shoulder_links" />
    <child link="elbow_link" />
  </joint>

  <joint name="gripper_joint" type="fixed">
    <origin xyz="0.01 0.37 -0.02" rpy="0 0 0" />
    <parent link="elbow_link" />
    <child link="gripper" />
  </joint>


  <!-- ros_control plugin-->
  <gazebo>
    <plugin name="gazebo_ros_control" filename= "libgazebo_ros_control.so">
      <robotNamespace>/</robotNamespace>
      <legacyModeNS>true</legacyModeNS>
    </plugin>
  </gazebo>
</robot>